### ISSUE TITLE
Import Waypoints from GPX File into Navigation

### ISSUE THREAD

**Issue by Sonwon1** [osmandapp/OsmAnd#20100]
### Describe the idea (required)

When importing a GPX file the import only shows the exact route. The imported file does not show waypoints between the start and end waypoints and does not provide proper turn-by-turn navigation. I propose a new GPX file import process type. When importing the file OsmAnd prompts the user for which import type the user wants: 

1) Exact route without waypoints and without turn-by-turn navigation (existing today).
2) Waypoint turn-by-turn navigation route that may not be the exact route (new).

Choice one would import the GPX files the same as OsmAnd currently works.

Choice two would add the the first 'wpt' entry in the GPX file into the navigation list as if the user was building a new navigation. Then the second and third and etc. 'wpt' entry in the GPX file into the navigation list. This would continue until the last 'wpt' was added to the navigation list. At this point the waypoint import process would be complete. The user could then edit the navigation list if desired and / or save the navigation for future use. Or just start navigating the route.

This would provide all of the waypoints with turn-by-turn navigation.

Many users of OsmAnd have complained about the waypoints in OsmAnd. This would fix the user's problems with missing waypoints and the lack of proper turn-by-turn instructions.

### Tell us about the expected behaviour (required)

I think above explains it all. Please let me know if there are questions.

### Tell us about alternatives you've considered (required)

I have tried importing numerous GPX files of various formats to obtain proper turn-by-turn instructions and to load the waypoints without success.

### Context (optional)

The current turn-by-turn instructions provided after importing a GPX file are not useful when driving on the Autobahn. For example, OsmAnd shows a right hand turn at the intersection of the Autobahn and does not prompt the user to get off the exit and does not draw the route using the exit. It is very confusing to know which way to go when there are two or more Autobahn intersections and exits to consider since there is only a right hand turn icon splashed down in the middle of the intersections. This caused me to miss my Autobahn exit causing me to drive 35 minutes out of my way to get back on route, more than once.

**Comment by yuriiurshuliak** [comment]
Could you share a specific GPX file for us to review?
Also have you tried attaching the track to the road? Does it impact turn-by-turn navigation in any way?

**Comment by 0709wiwiwi** [comment]
To follow up on the osm map matching operation, I take a look.

The Furkot export track files.

- The gpx header is incorrect. 
xsi:schemaLocation="http://www.topografix.com/GPX/1/1 http://www.topografix.com/GPX/1/1/gpx.xsd https://osmand.net">
- Correct is:
xsi:schemaLocation="http://www.topografix.com/GPX/1/1 http://www.topografix.com/GPX/1/1/gpx.xsd">

File 2024_Use_Hotels-_3.gpx
The track (path) quality is too poor. 
There is little to be done with that.
- Garbage in is therefore garbage out.

Compare with a Cruiser generated TBT navtrk.  Inclusive 30 TBT instructions attached in the trkpt. ( Routing by GH online)
[2024_Use_Hotels_-_3_wpt_navtrk by Cruiser.gpx.txt](https://github.com/user-attachments/files/15841070/2024_Use_Hotels_-_3_wpt_navtrk.by.Cruiser.gpx.txt)

OsmAnd imports these as regular track.  The OsmAnd Routing Based Java TBT generation is correct.

**Comment by Sonwon1** [comment]
@0709wiwiwi 

Let me dig into that, maybe I did something wrong with the export?

**Comment by 0709wiwiwi** [comment]
What I especially notice is that the track point definition in the Furkot version is much smaller and thus does not follow the roads very detailed and precise, which is what you should normally expect with a track. I think that the OsmAnd map matching tool does not handle that very well. The OsmAnd dev's can probably tell more.

**Comment by Sonwon1** [comment]
@0709wiwiwi 

When I create that route I think it was in low resolution. I will create a new file and try to upload it into Cruiser App. Does that app work with multiple waypoints?

**Comment by Sonwon1** [comment]
@0709wiwiwi 

I give up trying to create an importable gpx file. So if the routing service I am using cannot create importable gpx files that will provide turn-by-turn instructions then what web based routing service is a user suppose to use? I don't want to create routes on a small tablet screen and https://osmand.net/map is not ready.

**Comment by Sonwon1** [comment]
Thank you for the reply! I will look into these tools to see how I can use them for planning tours!

I like the routing service I am currently using since it allows planning a trip with multiple days in one go and breaks them up into multiple days for the navigation app. It also makes suggestions where the hotel stops would fall on the route. And allows to add time to see an attraction like 15 minutes for a photo stop or several hours for seeing a cave, etc. so the day does not become overly long.

I would like to stay with OsmAnd since it has most of the features I desire. Although the other navigation software you mentioned looks interesting too.

Between the routing and navigation service I will likely have to give up some features to make this all work together with the least amount of effort.

I notice when the gpx file import fails it is usually at roundabouts, freeway exits and intersections. And I have yet to test the new attach to roads algorithm. I read here that the beta has a speed bump so maybe it is usable now?

I thought of another test, will be back shortly.

**Comment by 0709wiwiwi** [comment]
Fine understood.

I see it this way. Each individual app definitely has features and tools that someone likes more or less. I don't see the range of apps as competitors either, but they can complement each other perfectly. What is mainly missing is a simple uniform file exchange method so that already added valuable elements such as TBT together with the via and shaping point info are not lost in exchanges. However, the exchange situation is currently as confusing unfortunately as it once was in Babylon.

**Comment by Sonwon1** [comment]
I have noticed that to be true!

**Comment by Sonwon1** [comment]
Since gps files are not standard and since importing them to create routes that provide turn-by-turn instructions is hit or miss, mostly miss. I believe having the option to load the route list with only <wpt> entries from the gps file would be a doable workaround. It is unlikely all of the gps files are cleaned up so they work with OsmAnd.

**Comment by 0709wiwiwi** [comment]
@Sonwon1 
It would sure be optimal if OsmAnd also offered you this simple [mode](https://youtu.be/tLNKzOjeYeE)

**Comment by vshcherb** [comment]
@yuriiurshuliak please summarize with screenshots I'm not sure I caught how exactly it would be easier to deal with GPX waypoints.

**Comment by yuriiurshuliak** [comment]
Issue Summary:

When importing a GPX file, OsmAnd only shows the start and end waypoints, missing intermediate waypoints and proper turn-by-turn navigation.

Steps to Reproduce:

- Import a GPX file with multiple waypoints.
- Observe that only the start and end points appear in the navigation list, missing all intermediate waypoints.

Conclusion:

When importing the tracks Furkot_Test_Route_-2.gpx and 2024_Use_Hotels-_3.gpx, waypoints do not display, although turn-by-turn instructions work correctly after attaching the track to the road.

Proposed Solution: 

New import option that includes all waypoints for navigation.

OsmAnd~ 4.8.0#2533m, released: 2024-06-19

<table style="width:100%">
  <tr>
<th><img src="https://github.com/osmandapp/OsmAnd/assets/127092082/4fb90865-b321-4bf3-bc71-a84f303283c2"  width="300" height="auto"/> </th>
 <th><img src="https://github.com/osmandapp/OsmAnd/assets/127092082/8acf1c3e-5ef7-46d1-bfe2-88f69f69ad72"  width="300" height="auto"/> </th>
  </tr>
</table>

<table style="width:100%">
  </tr>
  </tr>
  <tr>
    <td> <video source src="https://github.com/osmandapp/OsmAnd/assets/127092082/dfade637-f296-4ab8-beb4-0e007e406524"  width="300" height="auto"/>   </td>
    <td><video source src="https://github.com/osmandapp/OsmAnd/assets/127092082/071df91e-b787-4e8a-bdb8-11caa16e5c6e"  width="300" height="auto"/>   </td>
  </tr>
</table>

**Comment by Sonwon1** [comment]
@yuriiurshuliak 
I've been thinking about the post about and I am not sure if I was clear of what I am asking for. When I import a GPS file track or route I would like the option to select all waypoints and then OsmAnd router adds the route to follow. I don't want the route to come from the GPS file since that appears to cause turn-by-turn instruction issues in my experience.

Or if there is a way to trigger a route recalculation based on OsmAnd router for the whole route after import that would be great too.

And thank you for looking into this!

**Comment by 0709wiwiwi** [comment]
@Sonwon1.

I tested the road matching further anyway by using OsmAnd web.
Setting OsmAnd web Track road matching | Java based routing |(10m)|.

Your original Furkot track demo version: Until both roundabouts correct, then jumbled.
 - I checked that import track briefly and noticed a rather nasty error.
See the image with that track glitch return. Road matching can't handle this.
![Furkot demo trackglitch](https://github.com/osmandapp/OsmAnd/assets/54167201/0c5c934c-6cdf-49ac-8771-12af11727668)

I removed that nasty trackglitch from the original file.  Find the new file in attachment.
[Furkot_Test_Route_-_2_cleaned.gpx.txt](https://github.com/user-attachments/files/15916836/Furkot_Test_Route_-_2_cleaned.gpx.txt)

Import into OsmAnd web and the TBT are correct.
Conclusion:  Garbage in = Garbage out.

**Comment by Sonwon1** [comment]
@0709wiwiwi 

I saw that error in the track I had exported and corrected the routing to fix the problem. I did not noticed it until 2 days ago and we were pass that discussion. However I do not believe the longer route I created has that problem and it did not work. It appears to me that turn-by-turn instructions are not working with an imported GPS track file when you add Autobahns with over passes. I suspect sometimes the track points are place over an underpass/overpass intersection and the turn-by-turn routing gets confused as to which road you are on under or over road which causes the turn instruction errors. Maybe that is a bug that needs to be looked at to see if I am correct in my suspicion or maybe I am completely wrong. It would explain why the route only showed a right hand turn when I should have taken an exit.

Also, I did report back to the routing service the problems you pointed out in the GPS files above. I am looking forward to OsmAnd.net/map website being fully functional.

**Comment by 0709wiwiwi** [comment]
@Sonwon1 
Roger. 
A track glitch is mostly caused by a sloppily placed shaping point falsely positioned onto a side road. Normal tracks do not keep shaping points so the cause can hardly be retraced. In a gpx navigation track this valuable info is kept so such errors can more easily be detected and rectified. Also helpfull to detect errors is to check the turnlist for unplanned u-turns. Anyway in order to produce correct usefull quality tracks planner accuracy and discipline is most important.

Adding TBT to an already excellent quality track design requires road matching.
However a road matching router result may sligthly differ with the original built. 
It remains a bit of hocus pocus and than hope for a favourable useful outcome
With a gpx navigation track, road matching is unnecessary as the TBT are present.

**Comment by Sonwon1** [comment]
@0709wiwiwi 

I wasn't talking about shaping points positioned on a side road and in that case I would totally agree. 

I am referring to track points dropped throughout the route. In the GPS file I am not seeing any formatting to define if a track point is on top of an overpass or under an overpass. So I suspect the track point is dropped on top of overpasses. So when the routing algorithm tries to create a TBT it ends up giving instructions to get off the Autobahn to route to the track point on top of the overpass and then back onto the Autobahn. Perhaps this is a failure of the software creating the GPS file? This is not my area of expertise.

I tired to duplicate my scenario however I am having difficulty with getting track points to align on top of an overpass so any occurrence is random.

If I am correct I may have two different possible solutions that could fix to the algorithm to make it ignore track points when dropped on an overpass.

**Comment by 0709wiwiwi** [comment]
@Sonwon1.
I know what you mean. 
I can only conclude that the road matching behaviour is too erratic and unpredictable to arrive at a **robust stable usable result** without worries. Testing the road matching (a difficult task anyway ) usability was my intention but it in no way meets the expectations.  I am not wasting more testing time on it yet.

Your actual and original question was to use gpx waypoints to build a good reproducible TBT traject  with it.  
Let's talk about that.  I'll prepare some more, and come back to this later.  Nice weather cycle time first now.

**Comment by 0709wiwiwi** [comment]
@Sonwon1.

Better manually (re)compile track trajects to avoid bad TBT instructions. Place shaping_via routepoints away from road intersections. When selecting existing (poi) gpx waypoints as routepoins the optimal positioning is out of control. Non optimal positioned routepoints can produce short unexpected track tails inclusive false TBT and u-turns.

Such annoyances can be solved by a tool that avoid all this.
The (free) correction range to avoid this is 40m by default. 
An extra shaping point still allows to create short deviations in range.
RouteYou web offers +/- similar by a fixed 100m correction range. 

Save a calculated result in an easy to share gpx navigation track.
A compatible receiver app will offer an 100% identical TBT track.


A demo using cycle node waypoints.
- Example without correction. 
![near_via](https://github.com/osmandapp/OsmAnd/assets/54167201/32d1c57e-8b25-47be-877e-b515e1b5e136)

- Example with correction.
![near_via](https://github.com/osmandapp/OsmAnd/assets/54167201/29862768-1714-451e-8334-be1e87ecb71d)

**Comment by Sonwon1** [comment]
@0709wiwiwi 

I understand that and my refined routes avoid those turn around tails. My waypoints are placed as needed for stop locations and to shape a route. I got into that habit when I used Google Maps (GM) for routing. I needed shaping points and since GM only allows waypoints I had to use waypoints to shape a route. With the new routing software it allows for shaping points so my new routes only have beginning, stop locations and end waypoints. And I correct waypoint placement errors as needed. The longer route I posted was refined however I think you said the GPS file was poorly constructed. The sooner OsmAnd.net/maps is completed the better.

I am okay with tossing out the track points and using just waypoints to build a route. I can either import them (easy but not possible atm) from a GPS file or enter them manually (painful).

Where are we going with this?

**Comment by 0709wiwiwi** [comment]
Ah so, 
Google maps. Simple and up to date.
Osm-based apps offer neither of both.
- Simple: Simplify is desired.
- Up to date: Not possible.

Example:

A. 2024_Use_Hotels_-_3
The Google maps track version: trk. gpx
[GMaps2GPX-directions-cbee8897e6ea453590bb276c10a3eb98-0-1718979277376.gpx.txt](https://github.com/user-attachments/files/15942945/GMaps2GPX-directions-cbee8897e6ea453590bb276c10a3eb98-0-1718979277376.gpx.txt)
By dragging (shaping) Google maps only partially produces a similar track as the B33 is blocked.

B. Test_Route_-_2
The Gmaps track version: trk.gpx
[GMaps2GPX-directions-9140ce0003d149caa1fb2f36cb576e4e-0-1719124566349.gpx.txt](https://github.com/user-attachments/files/15942947/GMaps2GPX-directions-9140ce0003d149caa1fb2f36cb576e4e-0-1719124566349.gpx.txt)

A (manual) "enhanced" Gmaps track version: navtrk.gpx
[GMaps2GPX-directions-9140ce0003d149caa1fb2f36cb576e4e-0-1719124566349_navtrk.gpx.txt](https://github.com/user-attachments/files/15942948/GMaps2GPX-directions-9140ce0003d149caa1fb2f36cb576e4e-0-1719124566349_navtrk.gpx.txt)

Import > Navigate > Go.  (With the right app)
![Turncommands](https://github.com/osmandapp/OsmAnd/assets/54167201/84a51a8a-701c-43f0-ab86-59ee8dd4a209)

**Comment by Sonwon1** [comment]
Exactly, Google Maps is far far far from a perfect product.

**Comment by 0709wiwiwi** [comment]
Gmaps (car) works solidly in the provided app setup. Trajectory use in external apps is not encouraged. It is not "OpenGmaps".

**Comment by Sonwon1** [comment]
Check if you just want beginning to end with a few stops thrown in it is solid for that.

However this is getting way off topic and the thread is getting long.

I'll be pleased if waypoints import with the GPX file. There should also be a way to trigger a recalculate route too. Although I suppose you can do that now by faking another waypoint or route end or change one of the route restrictions like dirt roads, etc.

**Comment by 0709wiwiwi** [comment]
Pretty much anything you ask can already be done with that "other" app. Inserting pauses at via points is also coming.

**Comment by Sonwon1** [comment]
Not to drag this out, however the other apps are missing important pieces of functionality. After I build a tour route I then look through the route for interesting places to stop, see and take photos. Most routing services only have a subset of POI possibilities built into them and the result is limited. I have found three maybe four routing services that have more POIs than most. This is one reason why I am using a different routing service. Building a route on a small screen is tedious. If I cannot import GPS files then I have to manually transfer waypoints into OsmAnd as map markers from the GPS coordinates of the waypoint. Then I link those together forming a route that OsmAnd then builds into a mostly flawless turn-by-turn instructions. To say the least this is time consuming. I do not get a comprehensive list of POIs from most routing services. I hope this helps you to understand better where the problem lies in my usage of OsmAnd. Okay, I am all ears now so to read.

**Comment by 0709wiwiwi** [comment]
Find more bikers in the next forums. Ask how your colleagues handle this.
https://github.com/devemux86/cruiser/discussions
https://forum.kurviger.de/
To see what to expect find a single discourse.
https://github.com/devemux86/cruiser/discussions/209

**Comment by Sonwon1** [comment]
Thank you for the suggestions.

I've researched this at several websites with Kurviger being one of them. I was unaware of Cruiser until this thread.  I also went to https://advrider.com/f/ forums for more help where most adventure riders hang out.

**Comment by yuriiurshuliak** [comment]
Users have reported issues related to the import of the GPX file, where intermediate waypoints are being skipped, and incorrect Turn-by-Turn instructions.

Steps to Reproduce:
- Import the GPX track Furkot_Test_Route_-2.gpx into OsmAnd.
- Observe whether all waypoints in the GPX file are imported and displayed on the map.
- Check the Turn-by-Turn instructions to see if they align correctly with the intended route.

After importing the Furkot_Test_Route_-2.gpx track, I observed that three waypoints were successfully imported and displayed. However, there was no indication of any missing waypoints based on my inspection. It would be helpful to know which specific waypoints you found to be skipped.

Regarding Turn-by-Turn instructions, it’s important to attach the track to the road to generate accurate instructions. 

@Sonwon1 Could you specify which waypoints were missed during the import? How should the track look after a successful import? Are the Turn-by-Turn instructions incorrect even after the track is attached to the road?

<table style="width:100%">
  <tr>
<th><img src="https://github.com/user-attachments/assets/2311d32e-0d9d-49a0-b256-1bf089c5e438"  width="300" height="auto"/> </th>
 <th><img src="https://github.com/user-attachments/assets/65c9a9b8-0887-44e3-a94b-50c1ab5baf27"  width="300" height="auto"/> </th>
  </tr>
</table>

**Comment by Sonwon1** [comment]
I am on vacation and will respond when I return in 8 days or so.

**Comment by Sonwon1** [comment]
@yuriiurshuliak 

I am back from holiday, suffering from jet lag so I may make mistakes. thank you for being patient.

This request is asking for waypoints to be loaded to the navigation list so OsmAnd can build a new navigation route from the waypoints. During this special import all non waypoints should be ignored. I realize the route is likely to change from the original route. 

Your screenshot is from the import point list. After importing select navigation and you will not find the Unterbessenbach waypoint in the list.

When you load the Furkot_Test_Route_-_2.gpx.txt or any gpx file the waypoints are not loaded to the navigation waypoint list. The track/route is loaded however it does not provide proper navigation instructions. This happens with gpx track and route files.

The two examples of failures I remember after importing a gpx file are 1) when a point landed on an overpass OsmAnd navigation instructions would route off the Autobahn (freeway). So you would drive to the point on top of the overpass and then be routed back onto the Autobahn. 2) this can also occur where two or more Autobahns intersect causing all sorts of weird routing issues. If the waypoints were loaded to the navigation waypoint list then OsmAnd could rebuild the route without these problems since there would not be erroneous points on top of overpasses.

Suggestion for improving gpx file imports, all points in a gpx file that fall on an intersection (where two or more roads overlap) should not be loaded. I believe this could fix some of the problems with gpx file imports. Although this is for a different thread.

The Furkot_Test_Route_-_2.gpx.txt has three waypoints, start (Laufach), Unterbessenbach and End (Waldaschaff). 

The waypoint Unterbessenbach does show on the route drawn however it does not show in the navigation waypoint list.

Basically what is needed is a method to load only waypoints into the navigation list while ignoring all other points in the file and then trigger OsmAnd to build a route using the waypoints in the navigation list.

Maybe this is a bug?

**Comment by vshcherb** [comment]
Too complicated to understand
```
Suggestion for improving gpx file imports, all points in a gpx file that fall on an intersection (where two or more roads overlap) should not be loaded. I believe this could fix some of the problems with gpx file imports. Although this is for a different thread.
```

You can manually add points to navigation list and delete manually. We have plans to improve show along the route for Waypoints and Destination list

**Comment by Sonwon1** [comment]
> Too complicated to understand
> 
> ```
> Suggestion for improving gpx file imports, all points in a gpx file that fall on an intersection (where two or more roads overlap) should not be loaded. I believe this could fix some of the problems with gpx file imports. Although this is for a different thread.
> ```
> 
> You can manually add points to navigation list and delete manually. We have plans to improve show along the route for Waypoints and Destination list

What???

Manually adding points when you have 15-20 waypoints per day for a trip that spans over 14 days is ridiculously tedious. And not user friendly in my opinion.

Which parts are to complicated and I will try to rewrite in a clearer way? Or is it all of the noise in the thread that makes it appear to be complicated?

Currently OsmAnd is GPS track friendly however GPS routes are very unfriendly in my opinion. Also I am not the only user that has this opinion.

**Comment by vshcherb** [comment]
Then I think it's the same issue as this one https://github.com/osmandapp/OsmAnd/issues/20479

**Comment by Sonwon1** [comment]
> Then I think it's the same issue as this one #20479

@vshcherb I actually find #20479 confusing. Does copy waypoints from "any track" include GPS Route Files or just from a saved track stored in OsmAnd or both?

My understanding is GPS tracks and routes are not the same. Tracks are just track points to follow on a map without waypoints. Routes are a string of waypoints with track points between. Is this understanding incorrect?

I know OsmAnd will provide navigation instructions for both however the instructions are often less than ideal for the reasons I mentioned in this thread.

This is why I wanted a method to import just waypoints while ignoring the track points. So OsmAnd can build the track between the waypoints. This would avoid the problems I described in this thread.
